content/2022-02-25-pattern-matching-with-named-fields.md

```diff
@@ -0,0 +1,306 @@
+
+* Allow the same feature of case classes:
+  * Provide arbitrary names
+  * Enable [`@deprecatedName`](https://www.scala-lang.org/api/current/scala/deprecatedName.html), which leads to multiple names for a single field. Using both should be an error.
+* Maybe provide additional named fields. For example for our `User`:
+
+  ```scala
+  case User(ageInDays = 3650) => ...
+  ```
+
+  But this could be an awful feature.
+* Maybe leave fields without names
+
+This SIP proposes basically to use [shapeless records](https://github.com/milessabin/shapeless/wiki/Feature-overview:-shapeless-2.0.0#extensible-records) for the specific purpose of pattern matching.
+For incomplete list of alternatives see [alternative desugaring](#alternative-desugaring) below.
+
+Pro:
+
+* with shapeless records we have a good understanding what we are doing, however we need to tweak the encoding a bit
+* with some helper definitions the usage could be quite short
+* reuse possible
+
+Con:
+
+* opens an can of worms, as this encoding could be useful for other requested features like [named tuples](named-tuple), which would require more thought-out design
+* no support for `@deprecatedName`
+* identifiers are represented with string literals
+
+Example for user:
+
+```scala
+object User:
+  def unapply(user: User): Option[
+    (String, Int, String) & 
+    { type Names = ("name", "age", "city") }
+  ] = Some((user.name, user.age, user.city)).asInstanceOf
+```
+
+// TODO: Describe why Names can also be use toplevel.
+
+The reason to pull the names out, instead of keeping them near to their type, is to prevent extractors to accidentally leaking the name.
+
+## Implementation
+
+If a pattern with a name is encountered, the compiler looks up list of provided names and places the trees accordingly.
+
+The list of names either provided by the return type of the unapply method or by the constructor list of the case class.
+
+Example:
+
+```scala
+// a match clause like
+case User(<tree1>, city = <tree2>) => ???
+
+// gets rewritten to:
+case User(
+  <tree1>, // because tree1 is positional pattern
+  _, // because name isn't mentioned
+  <tree2>, // because city is the third parameter of user
+)
+```
+
+This allows us to glace over the details of how pattern matching gets desugared further down the line.
+
+## Drawbacks
+
+### Allow skipping arguments
+
+Whenever a single named argument is used in pattern, the pattern can have fewer arguments than the unapply provides. This is driven by the motivation the make pattern matching extensible.
+But this leads to (arguably small) inconsistency, as pointed out by Lionel Parreaux in the [Scala Contributors Thread](https://contributors.scala-lang.org/t/pattern-matching-with-named-fields/1829/44). This could lead users to use a named pattern, just to skip all parameters.
```

Comment:
This particular aspect troubles me. This will cause to silently ignore fields if we forget some.

I strongly believe we should have explicit syntax to deliberately ignore the remaining fields, if that's what we want. Perhaps something like
```scala
case User(name, city = c, _*) =>
```
with the generalized understanding that `_*` deliberately ignores what's left, whether it's single fields or variadic fields.

Reply:
@sjrd I did not understand what troubles you, normally one destructures to use the extracted value, so what is a case in which forgetting fields causes an issue?

Reply:
Basically, refactoring.

If I already have some `case User(name, city = c) =>` in my code, and then later I change the definition of `User` to add a new field, there is a pretty good chance that there are at least some `case User(...)` that will have to take this field into account. In many cases, for me, most if not all of my `case User` will need to take it into account. I want the compiler to help me with that refactoring. If we silently ignore any omitted field, refactoring becomes risky.